shadowsocks/stream.go

```diff
@@ -165,76 +166,91 @@ func (sr *shadowsocksReader) init() (err error) {
 	return nil
 }
 
-// ReadBlock reads and decrypts a single signed block of ciphertext.
-// The block will match the given decryptedBlockSize.
-// The returned slice is only valid until the next Read call.
-func (sr *shadowsocksReader) readBlock(decryptedBlockSize int) ([]byte, error) {
-	if err := sr.init(); err != nil {
-		return nil, err
-	}
-	cipherBlockSize := decryptedBlockSize + sr.aead.Overhead()
-	if cipherBlockSize > cap(sr.buf) {
-		return nil, io.ErrShortBuffer
-	}
-	buf := sr.buf[:cipherBlockSize]
+// readBlock reads and decrypts a single signed block of ciphertext.
+// The block must exactly match the size of `buf`, including overhead.
+// Returns an error only if the block could not be read.
+func (sr *shadowsocksReader) readBlock(buf []byte) error {
 	_, err := io.ReadFull(sr.reader, buf)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	buf, err = sr.aead.Open(buf[:0], sr.counter, buf, nil)
+	_, err = sr.aead.Open(buf[:0], sr.counter, buf, nil)
 	increment(sr.counter)
 	if err != nil {
-		return nil, fmt.Errorf("failed to decrypt: %v", err)
+		return fmt.Errorf("failed to decrypt: %v", err)
 	}
-	return buf, nil
+	return nil
 }
 
-func (sr *shadowsocksReader) Read(b []byte) (int, error) {
-	if err := sr.ensureLeftover(); err != nil {
+// ReadChunk returns the next decrypted chunk.  The caller must
+// complete its use of the returned buffer before the next call.
+func (sr *shadowsocksReader) ReadChunk() ([]byte, error) {
+	if err := sr.init(); err != nil {
+		return nil, err
+	}
+	sizeBuf := sr.buf[:2+sr.aead.Overhead()]
+	if err := sr.readBlock(sizeBuf); err != nil {
+		if err != io.EOF && err != io.ErrUnexpectedEOF {
+			err = fmt.Errorf("failed to read payload size: %v", err)
+		}
+		return nil, err
+	}
+	size := int(binary.BigEndian.Uint16(sizeBuf) & payloadSizeMask)
+	payloadBuf := sr.buf[:size+sr.aead.Overhead()]
```

Comment:
Not for this PR, but we could potentially reallocate the buffer here if we decide to use smaller buffers.

Reply:
Yeah, I'd like to enable optimizations like that here.